---
title: "Project 1"
author: "Kat Wiles"
date: "Friday, October 24, 2014"
output: html_document
---




###Introduction

When I started on this project, I originally wanted to work with some data from my lab. I struggled with the data processing, which took much more time than expcected. I then got a data set that had already been processed and tried playing around with the package [phyloseq](http://www.plosone.org/article/info%3Adoi%2F10.1371%2Fjournal.pone.0061217). Phyloseq is an R package that is used to visualize microbiome sample data. I had difficulty using the package with the data that I was given. For example, I was trying to make a relative abundance graph and I could not get the code to work. 

Instead of continuing with this project, I decided to scrap the whole thing and try to see if I could use the sample data on the [phyloseq website](http://joey711.github.io/phyloseq/) and reproduce many of the figures that are in the tutorial. The phyloseq package is designed to help scientists with reproducible data. I thought it would be interesting to see how reproducible their website is. This will also be useful way of learning how to use phyloseq. 

Once I learned how to phyloseq, I played around with the sample data to produce a few new figures and graphs. 


###Methods/Results

First I followed the [preprocessing tutorial](http://joey711.github.io/phyloseq/preprocess.html) with the GlobalPatterns data set within phyloseq. In which samples were removed if they were not seen more than 3 times in at least 20 of the samples. I also learned how to subset samples into varying taxonomic levels. 


```{r, echo=FALSE}
library("phyloseq") #load phyloseq library, version 1.10.0
data("GlobalPatterns") #loading Global Patterns data set
n.t<-ntaxa(GlobalPatterns) #exploring data: number of taxa
n.s<-nsamples(GlobalPatterns) #exploring data: number of samples
```

```{r, echo=FALSE}
#preprocessing
#to learn more about Globalpatterns data set see `?`(GlobalPatterns)
gpr  <- transform_sample_counts(GlobalPatterns, function(x) x / sum(x)) #transform data to relative abundance
gpfr <- filter_taxa(gpr, function(x) mean(x) > 1e-5, TRUE) # filter taxa so only OTUs mean greater than 10^-5kept
gp.ch <- subset_taxa(GlobalPatterns, Phylum=="Chlamydiae") # pulling out samples in phylum Chlamidiae
gp.ch <- prune_samples(sample_sums(gp.ch)>=20, gp.ch) #remove samples with >20 reads

gp.ch.merge <- merge_taxa(gp.ch, taxa_names(gp.ch)[1:5]) #merge first 5 samples in Chlamidiae set


gpsfbg <- subset_taxa(GlobalPatterns, Phylum =="Bacteroidetes") #pulling out Bacteriodetes #not in tutorial!!
gpsfbg <-  prune_samples(sample_sums(gpsfbg)>=20, gpsfbg) #removing samples with >20 reads #not in tutorial

gpsfbg <- tax_glom(gpsfbg, "Family") #mass Bacteriodetes only data set at rank of family
pt<- plot_tree(gpsfbg, color="SampleType", shape="Class", size="abundance") #create annotated tree

#graph not in tutorial, thus i left out of r markdown file

gp.ch.fa <-transform_sample_counts(gp.ch, function(OTU) OTU/sum(OTU) )#transforming abundance to functional abundance

gp <- filter_taxa(GlobalPatterns, function(x) sum(x > 3) > (0.2*length(x)), TRUE) #Remove taxa not seen more than 3 times in at least 20% of the samples. This protects against an OTU with small mean & trivially large C.V.

sample_data(gp)$human <- factor( get_variable(gp, "SampleType") %in% c("Feces", "Mock", "Skin", "Tongue") ) #define human vs non-human data and add to dataset

#standardizing abundances to mean sequence depth
total <- median(sample_sums(gp))
standf <- function(x, t=total) round(t * (x / sum(x)))
gps <- transform_sample_counts(gp, standf)
gpsf <- filter_taxa(gps, function(x) sd(x)/mean(x) > 3.0, TRUE) #Filter the taxa using a cutoff of 3.0 for the Coefficient of Variation

gpsfb <- subset_taxa(gpsf, Phylum=="Bacteroidetes") #Subset the data to Bacteroidetes, used in some plots
```


Next, I followed the [richness tutorial](http://joey711.github.io/phyloseq/plot_richness-examples.html). The first figure is plotting the richness of the OTUs in the Global Patterns data set. The graph is showing the different ways to measure alpha diversity. I added in color per sample type (NT) (not in tutorial). 


```{r, echo=FALSE, fig.height=6, fig.width=10}
#richness example

#setting up
library("ggplot2")
pv <- packageVersion("ggplot2") #version 1.0.0

theme_set(theme_bw())
pal = "Set1"
scale_colour_discrete <- function(palname = pal, ...) {
    scale_colour_brewer(palette = palname, ...)
}
scale_fill_discrete <- function(palname = pal, ...) {
    scale_fill_brewer(palette = palname, ...)
}


gp <- prune_taxa(taxa_sums(GlobalPatterns) > 0, GlobalPatterns) #prune OTUs not present
plot_richness(gp, color="SampleType")
```

Then, I looked at just the shannon graph (NT). 
```{r, echo=FALSE}
plot_richness(gp, color="SampleType", measures="Shannon")
```


Next I worked through the [ordination tutorial](http://joey711.github.io/phyloseq/plot_ordination-examples.html).

```{r, echo=FALSE}
#prepping data for ordination
theme_set(theme_bw())

#Remove OTUs that do not show appear more than 5 times in more than half the samples
gp <- GlobalPatterns 
wh0 <- genefilter_sample(gp, filterfun_sample(function(x) x > 5), A = 0.5 * nsamples(gp))
gp1 = prune_taxa(wh0, gp)

#Transform to even sampling depth.
gp1 <- transform_sample_counts(gp1, function(x) 1e+06 * x/sum(x))


#Keep only the most abundant five phyla.

phylum.sum <- tapply(taxa_sums(gp1), tax_table(gp1)[, "Phylum"], sum, na.rm = TRUE)
top5phyla <- names(sort(phylum.sum, TRUE))[1:5]
gp1 <- prune_taxa((tax_table(gp1)[, "Phylum"] %in% top5phyla), gp1)

#Define a human-associated versus non-human categorical variable:

human <- get_variable(gp1, "SampleType") %in% c("Feces", "Mock", "Skin", "Tongue")
sample_data(gp1)$human <- factor(human)
```


Samples and OTUs were graphed together in a biplot.


```{r, echo=FALSE}
GP.ord <- ordinate(gp1, "NMDS", "bray")
p1 = plot_ordination(gp1, GP.ord, type = "taxa", color = "Phylum", title = "taxa")
print(p1)


p1 + facet_wrap(~Phylum, 3)

p2 = plot_ordination(gp1, GP.ord, type = "samples", color = "SampleType", shape = "human")
p2 + geom_polygon(aes(fill = SampleType)) + geom_point(size = 5) + ggtitle("samples")

p3 = plot_ordination(gp1, GP.ord, type = "biplot", color = "SampleType", shape = "Phylum", 
    title = "biplot")
# Some stuff to modify the automatic shape scale
GP1.shape.names = get_taxa_unique(gp, "Phylum")
GP1.shape <- 15:(15 + length(GP1.shape.names) - 1)
names(GP1.shape) <- GP1.shape.names
GP1.shape["samples"] <- 16
p3 + scale_shape_manual(values = GP1.shape)
```

As you can see this brought up a suite of errors, even though I copied and pasted directly from the website, with no mention of these errors. With my unfamiliarity with the program, I was unable to fix the errors. There was mention fo a pch error, but this is not in the code for me to manipulate. 

Next I worked with [barplot tutorial](http://joey711.github.io/phyloseq/plot_bar-examples.html) to see if I could recreate the barplots. 

```{r, echo=FALSE}
par(mfrow=c(2,2))
#barplots
theme_set(theme_bw())
gp.ch = subset_taxa(GlobalPatterns, Phylum == "Chlamydiae") #pulling out subset in phylum Chlamydiae

plot_bar(gp.ch) #basic bar plot
plot_bar(gp.ch, fill="Genus") #expanding on bar plot to fill with the different genera
plot_bar(gp.ch, x="SampleType", fill="Genus") #group samples by where sample was from
plot_bar(gp.ch, "Family", fill="Genus", facet_grid=~SampleType) #using facets to separate sample types
```


Lastly, I worked with heat maps to see if the data could be recreated using the [heat map tutorial](http://joey711.github.io/phyloseq/plot_heatmap-examples.html).

```{r, echo=FALSE}
#heatmap
par(mfrow=c(1,3))

gpt <- subset_taxa(GlobalPatterns, Kingdom == "Bacteria")
gpt <- prune_taxa(names(sort(taxa_sums(gpt), TRUE)[1:300]), gpt)
plot_heatmap(gpt, sample.label = "SampleType")

gpac <- subset_taxa(GlobalPatterns, Phylum == "Crenarchaeota") #creating subset in phylum Crenarchaeota

plot_heatmap(gpac)
plot_heatmap(gpac, "NMDS", "jaccard") #heatmap with ordination

```


###Results

I was able to recreate most of the plots that were on the phyloseq tutorial website, and even created a few plots of my own. However, I had a lot of difficulty and errors with the ordination plots.


###Discussion

This project was based on reproducibility. I wanted to test if I could reproduce some of the tutorials on the phyloseq website, and hopefully expand upon the plots a little bit. I was initially hoping by doing so, I would become more familiar with the program and try again to make plots with my data. However, I do not feel the tutorial prepared me well enough to do so, just yet. 

when I had problems reproducing the ordination plots, I became frustrated because I could not figure out how to fix the plots. For example, I had a pch error. In the code visible to me, pch was not there. Hence, this brings up an issue with reproducible data when using someone eles's code. If all of the code is not visible, then it would be challenging to know what they did so you know how to fix it. It also makes it challenging to compare different versions of the code when the code is updated. 

For the most part, phyloseq tutorial is reproducible. However, for the novice, there is not enough information in the tutorial to teach you how to use the program for your own data. There was a lot of preprocessing (see code in beginning) that I spent a lot of time doing, but did not use most of it. 

In the future, I want learn how to use phyloseq and use it to analyze my data. There are a lot of features to import data from Mothur and Qiime that seems to be really useful. 







